File: benchmarks/array2d_tests/array2d.cpp

/*
Copyright (c) 2019 Oracle

The Universal Permissive License (UPL), Version 1.0

Subject to the condition set forth below, permission is hereby granted to any person
obtaining a copy of this software, associated documentation and/or data (collectively the
"Software"), free of charge and under any and all copyright rights in the Software, and
any and all patent rights owned or freely licensable by each licensor hereunder covering
either (i) the unmodified Software as contributed to or provided by such licensor, or (ii)
the Larger Works (as defined below), to deal in both

(a) the Software, and

(b) any piece of software and/or hardware listed in the lrgrwrks.txt file if one is
included with the Software (each a “Larger Work” to which the Software is contributed by
such licensors),

without restriction, including without limitation the rights to copy, create derivative
works of, display, perform, and distribute the Software and make, use, sell, offer for
sale, import, export, have made, and have sold the Software and the Larger Work(s), and to
sublicense the foregoing rights on either these or other terms.

This license is subject to the following condition:

The above copyright notice and either this complete permission notice or at a minimum a
reference to the UPL must be included in all copies or substantial portions of the
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.
*/


#include <iostream>
#include "Blackhole.hpp"

static const int SIZE = 64;

//@BenchmarkMode(Mode.AverageTime)
//@OutputTimeUnit(TimeUnit.NANOSECONDS)
//@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
//@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
//@Fork(1)
@State(Scope.Benchmark)
class Array2DTests {
public:

  @State(Scope.Benchmark)
  struct Arrays {
    int arr[SIZE][SIZE];
    int* arr2[SIZE][SIZE];
    //int (*arr3[SIZE])[SIZE];
    int* arr3[SIZE];
  };

  @Setup(Level.Trial)
  void setup(Arrays &arrs) {
    for (int i = 0; i < SIZE; ++i) {
      arrs.arr3[i] = new int[SIZE];
      for (int j = 0; j < SIZE; ++j) {
	arrs.arr[i][j] = i + j;
	arrs.arr2[i][j] = new int;
	*(arrs.arr2[i][j]) = i + j;
	//(*(arrs.arr3[i]))[j] = i + j;
	(arrs.arr3[i])[j] = i + j;
      }
    }

    std::cout << "Array size = " << SIZE << std::endl;
  }

private:

public:
   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int seq_by_rows(Arrays &arrs) {
     int sum = 0;
     
     for (int i = 0; i < SIZE; ++i) {
       for (int j = 0; j < SIZE; ++j) {
	 sum += arrs.arr[i][j];
       }
     }
     
     return sum;
   }

   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int seq_by_cols(Arrays &arrs) {
     int sum = 0;
     
     for (int j = 0; j < SIZE; ++j) {
       for (int i = 0; i < SIZE; ++i) {
	 sum += arrs.arr[i][j];
       }
     }
     
     return sum;
   }

   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int seq_by_rows_indirect(Arrays &arrs) {
     int sum = 0;
     
     for (int i = 0; i < SIZE; ++i) {
       for (int j = 0; j < SIZE; ++j) {
	 sum += *(arrs.arr2[i][j]);
       }
     }
     
     return sum;
   }

   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int seq_by_cols_indirect(Arrays &arrs) {
     int sum = 0;
     
     for (int j = 0; j < SIZE; ++j) {
       for (int i = 0; i < SIZE; ++i) {
	 sum += *(arrs.arr2[i][j]);
       }
     }
     
     return sum;
   }

   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int seq_by_rows_indrows(Arrays &arrs) {
     int sum = 0;
     
     for (int i = 0; i < SIZE; ++i) {
       for (int j = 0; j < SIZE; ++j) {
	 sum += (arrs.arr3[i])[j];
       }
     }
     
     return sum;
   }

   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.NANOSECONDS)
   @Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
   @Fork(1)
   @Benchmark
   static int seq_by_cols_indrows(Arrays &arrs) {
     int sum = 0;
     
     for (int j = 0; j < SIZE; ++j) {
       for (int i = 0; i < SIZE; ++i) {
	 sum += (arrs.arr3[i])[j];
       }
     }
     
     return sum;
   }

};


//int Array2DTests::InputValue::rep;
//volatile long Array2DTests::InputValue::number = 25;
//long Array2DTests::InputValue::number = 25;

File: benchmarks/array2d_tests/Array2D.java
/*
 * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

//package org.graalvm.benchmarks;
package leick;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Level;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Param;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
//import org.graalvm.polyglot.*;
import org.openjdk.jmh.annotations.TearDown;
import org.openjdk.jmh.annotations.Warmup;
import org.openjdk.jmh.infra.Blackhole;

//import java.io.File;
//import java.io.IOException;
import java.util.concurrent.TimeUnit;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Fork(1)
public class Array2D {

    static final int SIZE = 64;

    @State(Scope.Benchmark)
    public static class Arrays {
	int[][] arr = new int[SIZE][SIZE];
	
	@Setup(Level.Iteration)
        public void setup() {
	    for (int i = 0; i < SIZE; i++) {
		for (int j = 0; j < SIZE; j++) {
		    arr[i][j] = i + j;
		}
	    }
	}
    };


   @Benchmark
   public int seq_by_rows(Arrays arrs) {
       int sum = 0;
       
       //for (int i = 0; i < SIZE; i++) {
       //	   for (int j = 0; j < SIZE; j++) {
       //	       assert i < arrs.arr.length : " seq_by_rows, i oob";
       //	       assert j < arrs.arr[i].length : " seq_by_rows, j oob";
       //	       sum += arrs.arr[i][j];
       //	   }
       //}
       
       for (int i = 0; i < arrs.arr.length; i++) {
       	   for (int j = 0; j < arrs.arr[i].length; j++) {
       	       sum += arrs.arr[i][j];
       	   }
       }
       
       return sum;
   }

   @Benchmark
   public int seq_by_cols(Arrays arrs) {
       int sum = 0;
       
       //for (int j = 0; j < SIZE; j++) {
       //	   for (int i = 0; i < SIZE; i++) {
       //	       assert i < arrs.arr.length : " seq_by_cols, i oob";
       //	       assert j < arrs.arr[i].length : " seq_by_cols, j oob";
       //	       sum += arrs.arr[i][j];
       //	   }
       //}
       
       for (int j = 0; j < arrs.arr[0].length; j++) {
       	   for (int i = 0; i < arrs.arr.length; i++) {
       	       sum += arrs.arr[i][j];
       	   }
       }
       
       return sum;
   }

}

